src/rate_limit.rs

```diff
@@ -0,0 +1,239 @@
+        call_frequency: Duration,
+        clock: C,
+    ) -> Self {
+        let mut peer_map: HashMap<PublicKey, PeerRecord> = HashMap::new();
+        for peer in peers {
+            peer_map.insert(peer, PeerRecord::new(true, call_count));
+        }
+
+        let last_update = clock.now();
+        Self {
+            peer_map,
+            clock,
+            call_count,
+            call_frequency,
+            last_update,
+        }
+    }
+
+    fn needs_update(&self) -> bool {
+        self.clock.now().duration_since(self.last_update) >= self.call_frequency
+    }
+
+    fn update(&mut self) {
+        // We can safely delete offline peers because they would have their call count updated at
+        // this point anyway. We want to delete so that we don't allow an infinitely growing queue.
+        self.peer_map.retain(|_, v| v.online);
+
+        // Refresh allowed call counts per-peer that's left online.
+        for (_, v) in self.peer_map.iter_mut() {
+            v.remaining_calls = self.call_count;
+        }
+
+        self.last_update = self.clock.now();
+    }
+
+    fn hit(&mut self, peer_key: PublicKey) -> bool {
+        match self.peer_map.get_mut(&peer_key) {
+            Some(v) => {
+                if v.remaining_calls == 0 {
+                    return false;
+                }
+
+                v.remaining_calls -= 1;
+                true
+            }
+            None => false,
+        }
+    }
+}
+
+impl<C: Clock> RateLimiter for TokenLimiter<C> {
+    fn peer_connected(&mut self, peer_key: PublicKey) {
+        self.peer_map
+            .entry(peer_key)
+            .and_modify(|e| e.online = true)
+            .or_insert(PeerRecord::new(true, self.call_count));
+    }
+
+    fn peer_disconnected(&mut self, peer_key: PublicKey) {
+        self.peer_map
+            .entry(peer_key)
+            .and_modify(|e| e.online = false);
+    }
+
+    fn peers(&self) -> Vec<PublicKey> {
+        self.peer_map
+            .clone()
+            .into_iter()
+            .filter(|p| p.1.online)
+            .map(|p| p.0)
+            .collect::<Vec<PublicKey>>()
+    }
+
+    fn query_peer(&mut self, peer_key: PublicKey) -> bool {
+        if self.needs_update() {
+            self.update();
+        };
+
+        self.hit(peer_key)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::tests::test_utils::pubkey;
+    use core::ops::SubAssign;
+    use mockall::mock;
+    use tokio::time::{Duration, Instant};
+
+    const TEST_COUNT: u8 = 2;
+    const TEST_FREQUENCY: Duration = Duration::from_secs(1);
+
+    mock! {
+        FixedClock{}
+
+        impl Clock for FixedClock{
+            fn now(&self) -> Instant;
+        }
+    }
+
+    #[test]
+    fn test_peer_connection() {
+        let pk_0 = pubkey(0);
+        let pk_1 = pubkey(1);
+
+        let mut clock = MockFixedClock::new();
+        let start_time = Instant::now();
+        clock.expect_now().returning(move || start_time);
```

Comment:
instead of defining start_time above and using "move" could we just define it inside returning?